test/widgets/action_sheet_test.dart

```diff
@@ -99,6 +103,302 @@ void main() {
+
+    /// Prepare store and bring up a topic action sheet.
+    ///
+    /// If `isChannelMuted` is `null`, the user is not subscribed to the
+    /// channel.
+    Future<void> setupToTopicActionSheet(WidgetTester tester, {
+      required bool? isChannelMuted,
+      required UserTopicVisibilityPolicy visibilityPolicy,
+      int? featureLevel,
+    }) async {
+      addTearDown(testBinding.reset);
+
+      channel = eg.stream();
+      topic = 'isChannelMuted: $isChannelMuted, policy: $visibilityPolicy';
+
+      final account = eg.selfAccount.copyWith(zulipFeatureLevel: featureLevel);
+      final subscriptions = isChannelMuted == null ? <Subscription>[]
+        : [eg.subscription(channel, isMuted: isChannelMuted)];
+      await testBinding.globalStore.add(account, eg.initialSnapshot(
+        realmUsers: [eg.selfUser],
+        streams: [channel],
+        subscriptions: subscriptions,
+        userTopics: [eg.userTopicItem(channel, topic, visibilityPolicy)],
+        zulipFeatureLevel: featureLevel));
+      store = await testBinding.globalStore.perAccount(account.id);
+      connection = store.connection as FakeApiConnection;
+
+      connection.prepare(json: eg.newestGetMessagesResult(
+        foundOldest: true, messages: [
+          eg.streamMessage(stream: channel, topic: topic)]).toJson());
+      await tester.pumpWidget(TestZulipApp(accountId: account.id,
+        child: MessageListPage(
+          initNarrow: TopicNarrow(channel.streamId, topic))));
+      await tester.pumpAndSettle();
+
+      await tester.longPress(find.descendant(
+        of: find.byType(RecipientHeader), matching: find.text(topic)));
+      // sheet appears onscreen; default duration of bottom-sheet enter animation
+      await tester.pump(const Duration(milliseconds: 250));
+    }
+
+    void checkButtons(List<Finder> expectedButtonFinders) {
+      if (expectedButtonFinders.isEmpty) {
+        check(find.byType(BottomSheet)).findsNothing();
+        return;
+      }
+      check(find.byType(BottomSheet)).findsOne();
+
+      for (final buttonFinder in expectedButtonFinders) {
+        check(buttonFinder).findsOne();
+      }
+      check(find.bySubtype<UserTopicUpdateButton>())
+        .findsExactly(expectedButtonFinders.length);
+    }
+
+    void checkUpdateUserTopicRequest(UserTopicVisibilityPolicy expectedPolicy) async {
+      check(connection.lastRequest).isA<http.Request>()
+        ..url.path.equals('/api/v1/user_topics')
+        ..bodyFields.deepEquals({
+          'stream_id': '${channel.streamId}',
+          'topic': topic,
+          'visibility_policy': '${expectedPolicy.toJson()}',
+        });
+    }
+
+    testWidgets('unmuteInMutedChannel', (tester) async {
+      await setupToTopicActionSheet(tester,
+        isChannelMuted: true,
+        visibilityPolicy: UserTopicVisibilityPolicy.none);
+      await tester.tap(unmuteInMutedChannel);
+      await tester.pump();
+      checkUpdateUserTopicRequest(UserTopicVisibilityPolicy.unmuted);
```

Comment:
… how about just:
```suggestion
      await tester.tap(find.text('Unmute topic'));
      await tester.pump();
      checkUpdateUserTopicRequest(UserTopicVisibilityPolicy.unmuted);
```

I think that covers all the facts that matter to the app's behavior as far as the user is concerned. And it both makes the test simpler, and less brittle to changes in the implementation.

Reply:
(Well, I guess there's one more fact this version doesn't check: the icon. But the icon isn't really being checked in the current revision either — just some inputs toward computing the icon.

It'd be a nice bonus to also check what icon is shown for each button, but I'll be happy merging this without that check.)

Reply:
Maybe something like:
```dart
    final mute =     find.text('Mute topic');
    final unmute =   find.text('Unmute topic');
    final follow =   find.text('Follow topic');
    final unfollow = find.text('Unfollow topic');
```

so that the later tests like this one:
```dart
      final testCases = [
        (false, UserTopicVisibilityPolicy.muted,    [unmute, follow]),
        (false, UserTopicVisibilityPolicy.none,     [mute, follow]),
        (false, UserTopicVisibilityPolicy.unmuted,  [mute, follow]),
        (false, UserTopicVisibilityPolicy.followed, [mute, unfollow]),

        (true,  UserTopicVisibilityPolicy.muted,    [unmute, follow]),
        (true,  UserTopicVisibilityPolicy.none,     [unmute, follow]),
        (true,  UserTopicVisibilityPolicy.unmuted,  [mute, follow]),
        (true,  UserTopicVisibilityPolicy.followed, [mute, unfollow]),

        (null,  UserTopicVisibilityPolicy.none,     <Finder>[]),
      ];
```
can still be formatted compactly.

Reply:
Sure, that works.